protocol/directory.go

```diff
@@ -0,0 +1,48 @@
+package protocol
+
+import (
+	"log"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ConiksUserPolicies struct {
+	Username               string
+	AllowUnsignedKeychange bool
+	AllowPublicLookup      bool
+}
+
+type ConiksDirectory struct {
+	*merkletree.PAD
+}
+
+func InitDirectory(policies merkletree.Policies, signKey sign.PrivateKey,
+	dirSize uint64) *ConiksDirectory {
+	pad, err := merkletree.NewPAD(policies, signKey, dirSize)
+	if err != nil {
+		panic(err)
+	}
+	return &ConiksDirectory{pad}
+}
+
+func (d *ConiksDirectory) Register(uname string, key []byte) (
+	*merkletree.AuthenticationPath, *merkletree.TemporaryBinding, ErrorCode) {
+	// check whether the name already exists
+	// in the directory before we register
+	ap, err := d.Lookup(uname)
+	if err != nil {
+		return nil, nil, ErrorInternalServer
+	}
+	if ap.Leaf.Value() != nil {
```

Comment:
I'm pretty sure about it. We're doing it in our tests, e.g., https://github.com/coniks-sys/coniks-go/blob/master/merkletree/pad_test.go#L88-L92